tidy3d/components/source.py

```diff
@@ -291,7 +301,72 @@ def amp_time(self, time: float) -> complex:
         return const * offset * oscillation * amp
 
 
-SourceTimeType = Union[GaussianPulse, ContinuousWave]
+class CustomSourceTime(Pulse):
+    """Custom source time dependence.
+
+    Note
+    ----
+    To interpolate to simulation time steps, the value provided at the nearest time is used.
+
+    Example
+    -------
+    >>> sourcetime = CustomSourceTime.from_numpy(values=np.linspace(0, 1, 100), dt=.1)
+    """
+
+    sourcetime_dataset: Optional[TimeDataset] = pydantic.Field(
+        ..., title="Source time dataset", description="Dataset for storing the custom source time."
+    )
+
+    _sourcetime_dataset_none_warning = warn_if_dataset_none("sourcetime_dataset")
+
+    @classmethod
+    def from_numpy(
```

Comment:
maybe `from_values` or something makes more sense here? One could imagine `from_numpy` perhaps accepting the time coordinates instead of dt, for example.